pkg/server/server.go

```diff
@@ -222,15 +246,161 @@ func (s *Server) Run(ctx context.Context) error {
+
+	workspaceLister := kcpSharedInformerFactory.Tenancy().V1alpha1().Workspaces().Lister()
+
+	serverOptions.APIExtensionsNewSharedInformerFactoryFunc = func(client apiextensionsclient.Interface, resyncPeriod time.Duration) crdexternalversions.SharedInformerFactory {
+		f := apiextensionsinformers.NewSharedInformerFactory(client, resyncPeriod)
+		return &kcpAPIExtensionsSharedInformerFactory{
+			SharedInformerFactory: f,
+			workspaceLister:       workspaceLister,
+		}
+	}
+
 	cpOptions, err := genericcontrolplane.Complete(serverOptions)
 	if err != nil {
 		return err
 	}
 
-	server, err := genericcontrolplane.CreateServerChain(cpOptions, ctx.Done())
+	serverChain, err := genericcontrolplane.CreateServerChain(cpOptions, ctx.Done())
 	if err != nil {
 		return err
 	}
+	server := serverChain.MiniAggregator.GenericAPIServer
+
+	serverChain.GenericControlPlane.GenericAPIServer.Handler.GoRestfulContainer.Filter(func(req *restful.Request, res *restful.Response, chain *restful.FilterChain) {
+		ctx := req.Request.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewInternalError(fmt.Errorf("no RequestInfo found in the context")),
+				// TODO is this the right Codecs?
+				scheme.Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, res.ResponseWriter, req.Request,
+			)
+			return
+		}
+
+		// If it's not the core group, pass through
+		if requestInfo.APIGroup != "" {
+			chain.ProcessFilter(req, res)
+			return
+		}
+
+		// Special handling of the core ("") group
+
+		if requestInfo.IsResourceRequest {
+			// This is a CRUD request for somethig like pods. Try to see if there is a CRD for the resource. If so, let the CRD
+			// server handle it.
+			crdName := requestInfo.Resource + ".core"
+			_, err := serverChain.CustomResourceDefinitions.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().GetWithContext(ctx, crdName)
+			if err == nil {
+				serverChain.CustomResourceDefinitions.GenericAPIServer.Handler.NonGoRestfulMux.ServeHTTP(res.ResponseWriter, req.Request)
+				return
+			}
+		} else if req.Request.URL.Path == "/api/v1" || req.Request.URL.Path == "/api/v1/" {
+			// This is a discovery request. We may need to combine discovery from the GenericControlPlane (which has built-in v1 types)
+			// and CRDs, if there are any v1 CRDs.
+
+			// If we are retrieving the GenericControlPlane's v1 APIResources, let that request through.
+			const passthroughHeader = "X-Kcp-Api-V1-Discovery-Passthrough"
+			if _, passthrough := req.Request.Header[passthroughHeader]; passthrough {
+				chain.ProcessFilter(req, res)
+				return
+			}
+
+			// Get all the CRDs (in the context's logical cluster) to see if any of them are in v1
+			crds, err := serverChain.CustomResourceDefinitions.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().ListWithContext(ctx, labels.Everything())
+			if err != nil {
+				// FIXME
```

Comment:
No error handling here (yet)... does it just make sense to log and still return a response (the built in /api/v1 discovery)?

Reply:
can this even fail?